lib/spack/spack/multimethod.py

```diff
@@ -123,33 +123,55 @@ def _get_method_by_spec(self, spec):
         return self.default or None
 
     def __call__(self, package_self, *args, **kwargs):
+        """Execute the first method with a spec that matches the
+           package's spec or, if none, the default.
+
+           Raises a NoSuchMethodError if there is no default.
+        """
+        pkg_arg = args[0] if len(args) > 0 else None
+        spec_method = self.get_first_matching_method(package_self, pkg_arg)
+        if spec_method:
+            # If the package isn't being provided directly, it will be the
+            # first argument.
+            if package_self:
+                pkg = package_self
+            else:
+                pkg = pkg_arg
+                args = args[1:] if len(args) > 1 else set()
+            return spec_method(pkg, *args, **kwargs)
+
+    def get_first_matching_method(self, package_self, pkg_arg):
         """Find the first method with a spec that matches the
            package's spec.  If none is found, call the default
            or if there is none, then raise a NoSuchMethodError.
         """
-        spec_method = self._get_method_by_spec(package_self.spec)
+        # Allow for cases where __func__ is such that package_self is
+        # None and the package is passed as the first entry in args
+        # (e.g., stand-alone package testing).
+        pkg = package_self if package_self else pkg_arg
```

Comment:
@becker33 @tldahlgren What would be an example of a method with `package_self == None`? I think I don't understand what we refer to "stand-alone package testing".

Reply:
🤔